services/QuillLMS/app/services/skills_report/user_responses_processor.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+module SkillsReport
+  class UserResponsesProcessor < ApplicationService
+    EXPOSURE_HIGH_LIMIT = 15
+    EXPOSURE_MEDIUM_LIMIT = 6
+
+    attr_reader :responses, :name, :questions_hash
+
+    Result = Data.define(:name, :ema, :ema_previous, :trend, :exposure, :unique_answered, :replays_answered, :total_answered, :replay_rate, :last_practiced, :scores, :timespent)
+
+    delegate :ema, :ema_previous, :trend, to: :ema_trends, allow_nil: true
+
+    def initialize(responses:, name:, questions_hash:)
+      @responses = responses
+      @name = name
+      @questions_hash = questions_hash
+    end
+
+    def run
+      Result.new(name:, ema:, ema_previous:, trend:, exposure:, unique_answered:, replays_answered:, total_answered:, replay_rate:, last_practiced:, scores:, timespent:)
+    end
+
+    private def non_empty_responses = @non_empty_responses ||= responses.filter { |r| r.activity_session_id.present? && r.question_uid.present? }
+    private def unique_answered = @unique_answered ||= non_empty_responses.uniq { |r| r.question_uid }.count
+    private def total_answered = @total_answered ||= non_empty_responses.count
+    private def replays_answered = total_answered - unique_answered
+    private def latest_response = @latest_response ||= non_empty_responses.first
+    private def last_practiced = latest_response&.activity_session_completed_at ? Date.parse(latest_response.activity_session_completed_at) : nil
+    private def exposure
+      if total_answered >= EXPOSURE_HIGH_LIMIT
+        'High'
+      else
+        total_answered >= EXPOSURE_MEDIUM_LIMIT ? 'Medium' : 'Low'
+      end
+    end
+    private def replay_rate = (replays_answered / [total_answered.to_f, 1].max) * 100
+
+    private def ema_trends = @ema_trends ||= EmaTrendCalculator.run(scores:)
+    private def scores = @scores ||= questions_to_score.map { |response| QuestionResponseScorer.run(response:, question: questions_hash[response.question_uid].first) }
+
+    private def timespent
+      activity_responses = responses
+        .uniq(&:activity_session_id)
+
+      activity_responses
+        .map { |response| time_spent_for_response(response) }
+        .sum / (activity_responses.empty? ? 1 : activity_responses.length)
+    end
+
+    private def time_spent_for_response(response)
+      skill_count = Activity
+        .joins(:skill_groups)
+        .where(id: response.activity_id)
+        .count
+
+      return response.timespent unless skill_count > 0
+
+      response.timespent.to_f / skill_count
+    end
```

Comment:
This logic has known deviations from current spec that have to do with calculating `skill_count`, but the fundamental shape of this approach is likely to remain similar once that is fixed.